ENH: always make ndarrays from msgpack writable

llllllllll opened the pull request:
Addresses the case where 'compress' was not none. The old implementation
would decompress the data and then call np.frombuffer on a bytes
object. Because a bytes object is not a mutable buffer, the resulting
ndarray had writeable=False. The new implementation ensures that the
pandas is the only owner of this new buffer and then sets it to mutable
without copying it. This means that we don't need to do a copy of the
data coming in AND we can mutate it later. If we are not the only owner
of this data then we just copy it with np.fromstring.

Commit:
ENH: always make ndarrays from msgpack writable

Addresses the case where 'compress' was not none. The old implementation
would decompress the data and then call np.frombuffer on a bytes
object. Because a bytes object is not a mutable buffer, the resulting
ndarray had writeable=False. The new implementation ensures that the
pandas is the only owner of this new buffer and then sets it to mutable
without copying it. This means that we don't need to do a copy of the
data coming in AND we can mutate it later. If we are not the only owner
of this data then we just copy it with np.fromstring.

jreback commented on pandas/io/packers.py:
```diff
@@ -255,17 +261,36 @@ def unconvert(values, dtype, compress=None):
+            raise ValueError("compress must be one of 'zlib' or 'blosc'")
+
+        values = decompress(values)
+        if sys.getrefcount(values) == 2:
```
where did you get this beauty?

llllllllll replied:
I was just playing around with frombuffer and thinking of ways to avoid the copy here.

jreback replied:
Why can't we *always* just use ``np.frombuffer(buffer(values), dtype=....)``?

then just set the writeable flag?

llllllllll replied:
The issue ius that this we will be working with memory allocated for a `bytes`. For example:

```python
In [1]: buf = bytes(range(4))

In [2]: np.frombuffer(buf, dtype='uint8')
Out[2]: array([0, 1, 2, 3], dtype=uint8)

In [3]: _2.flags.writeable = True

In [4]: _2[0] = 5

In [5]: buf
Out[5]: b'\x05\x01\x02\x03'
```

The only case where we can just use the buffer is when we are the sole owner of values and no one can reference `values` again. In the other cases we need to just copy the data so we can mutate while preserving normal python semantics of immutable strings.

jreback replied:
I c. Is this really a bottleneck, e.g. the copying? Does the compression help?

llllllllll replied:
The copying might be restrictive if you are unpacking a large dataframe because your high watermark is now 2n. We are unpacking data on a shared box so I would like to keep the memory usage low if possible. I also imagine that users who are using compression are recieving a large amount of data.

llllllllll replied:
I actually thing the ideal fix would be for blosc and zlib to have unpack functions that give back bytearrays so that we can do this in a more safe way. I will investigate if this is already available.

llllllllll replied:
it looks like there is no good way to get the compressors to return bytearrays in their current state.

llllllllll replied:
The implementation proposed in this PR is scary so I tried to solve this problem upstream in  [blosc](https://github.com/Blosc/python-blosc/pull/107) and [zlib](http://bugs.python.org/issue26379)

Hopefully we can swap to the more supported solutions when available and only use this hack when we need to.

jreback commented:
cc @kawochen

jreback commented:
cc @kawochen

jreback commented on pandas/io/tests/test_packers.py:
```diff
@@ -57,6 +59,20 @@ def check_arbitrary(a, b):
         assert(a == b)
 
 
+@contextmanager
```
isnt't this like a ``unittest.Mock`` object? (well what you are doing below)

llllllllll replied:
I thought that was only added to the stdlib in py3

Commit:
TST: adds test for memoized empty string and chars

In many capi functions for python the empty string and single characters
are memoized to return preallocated bytes objects. This could be a
potential problem with the implementation of pandas.io.packers.unconvert
so we are adding a test for this explicitly.

Currently neither zlib nor blosc hit this case because they use
PyBytes_FromStringAndSize(NULL, n) which does not pull from the shared
pool of bytes objects; however, this test will guard us against changes
to this implementation detail in the future.

kawochen commented:
python magic!  there might be performance warnings during debugging when the debugger holds more references

llllllllll commented:
I think that warning while debugging is unfortunate but correct. Do you think that it would be worth amending the warning to add "or you are in a debugger"?

kawochen commented:
I'm not sure what the best thing to do is.  In general I believe relying on ref count is sketchy (e.g. count could be incremented between function calls).  But then I think `zlib` doesn't really return anything that's shared, even for small strings (I didn't manage to get any performance warnings with crafted cases).

I'm only minimally familiar with the C-API.  Is it possible that, although the ref count is 2, the string is interned during the query (maybe by `getrefcount` itself) and never destroyed, so subsequent attempts to use that string give unexpected results?

In moving data around, this `n` vs `2n` memory issue is very common, but I think it is a legit concern, and in this case it does seem a bit wasteful.

llllllllll commented:
Subsequent accesses to the `bytes` object will have unexpected results. Think of this as a `move` operation. I have checked the source for zlib and blosc to confirm that in their current state they will not be interning the strings; however, the test cases are to guard against changes to the implementation of these libraries. I also have open PRs upstream to both CPython and blosc that will make these functions return mutable buffers. We will need to support this hack for some time (especially with zlib) but at least there will be a more supported code path. I have an idea for a function that will factor out this logic a bit and ensure that users cannot get a reference back to the invalid bytes object. Right now this object is still exposed as `df._data.blocks[n].values.base.base` which is not ideal but also pretty deeply hidden/

Commit:
ENH: updates unconvert no-copy code to be safer

Pulls the logic that does the buffer management into a new private
function named `_move_into_mutable_buffer`. The function acts like a
move constructor from 'bytes' to `memoryview`. We now only call
`np.frombuffer` on the resulting `memoryview` object. This accomplishes
the original task of getting a mutable array from a `bytes` object
without copying with the added safety of making the `base` field of the
array that is returned to the user a mutable `memoryview`. This
eliminates the risk of users obtaining the original `bytes` object which
is now pointing to a mutable buffer and violating the immutability
contract of `bytes` objects.

llllllllll commented:
@kawochen I am updating the PR with a new implementation that I think is both safer and faster (or at least the same speed).

With this new implementation:

```python
In [34]: pd.read_msgpack(df.to_msgpack(compress='blosc'))._data.blocks[0].values.base.base
Out[34]: <memory at 0x7f71f8e81750>
```

However with the old implementation the base was the `bytes` object that shared a buffer with the array.

Here are profiling runs to show the differences:

These cells are used for all three profiling runs.
```python
In [1]: df = pd.DataFrame({'a': np.arange(1000000), 'b': np.arange(1000000)})

In [2]: packed = df.to_msgpack(compress='blosc')

Using `np.fromstring` to copy the data (this is just shown to make it clear why we cannot do this):

```python
In [3]: %timeit pd.read_msgpack(packed)
10 loops, best of 3: 168 ms per loop
```

Old implementation:

```python
In [3]: %timeit pd.read_msgpack(packed)
100 loops, best of 3: 3.06 ms per loop
```

Updated implementation:

```python
In [5]: %timeit pd.read_msgpack(packed)
100 loops, best of 3: 2.68 ms per loop
```

jreback commented on pandas/io/packers.py:
```diff
@@ -240,6 +246,77 @@ def convert(values):
+
+    Notes
+    -----
+    If you want to use this function you are probably wrong.
```
gr8 comment!

jreback commented on pandas/io/tests/test_packers.py:
```diff
@@ -57,6 +59,20 @@ def check_arbitrary(a, b):
+def patch(ob, attr, value):
+    noattr = object()  # mark that the attribute never existed
+    old = getattr(ob, attr, noattr)
+    setattr(ob, attr, value)
```
let's put this in ``pandas.utils.testing`` (also add a doc-string), basically pointing to a compat for Mock.

llllllllll commented:
Ugh, I forgot that `memoryview` has an `obj` attribute which is exactly like `ndarray`s `base` attribute. I am investigating ways to drop this in pure python but I am starting to think that we might just need to use a little ctypes to kill this object.

jreback commented:
This is on master (not  your branch)

```
# zlib
[22]: %timeit pd.read_msgpack('test.pak')
10 loops, best of 3: 144 ms per loop

In [23]: %timeit pd.read_pickle('test.pkl')
100 loops, best of 3: 13.7 ms per loop

# blosc
In [3]: %timeit pd.read_msgpack('test.pak')
100 loops, best of 3: 3.19 ms per loop
```

blosc >> zlib

Commit:
ENH: reimplement `_move_into_mutable_buffer` in C.

By writing our move function in C we can hide the original bytes object
from the user while still ensuring that the lifetime is managed
correctly. This implementation is designed to make it impossible to get
access to the invalid bytes object from pure python.

Commit:
TST: adds `pandas.util.testing.patch`

This function temporarily replaces a an attribute on an object for the
duration of some context manager. This is useful for patching methods to
inject assertions or mock out expensive operations.

jreback commented:
@llllllllll 

This is on 0.17.1
```
In [1]: df = pd.DataFrame({'a': np.arange(1000000), 'b': np.arange(1000000)})

In [2]: In [2]: packed = df.to_msgpack(compress='blosc')

In [3]: %timeit pd.read_msgpack(packed)
100 loops, best of 3: 4.97 ms per loop
```

How did you get the above?

jreback commented:
I am amazed on how fast ``blosc`` is. we should turn this on by default if ``blosc`` is available.

llllllllll commented:
sorry, the really slow run was what would happen if we copied the data after doing `blosc.decompress`. This code was never on master I just wanted to show that it was worthwhile to use a more complex approach to avoid this copy.

llllllllll commented:
ping, how do you feel about this?

kawochen commented:
no trace of `bytes` left :+1:

llllllllll commented:
hype

jreback commented:
what triggers the error on master?

llllllllll commented:
which error?

jreback commented:
the problem IIUC which you are solving is that you are getting a non writable buffer back right?
as there doesn't seem to be a perf issue with master

llllllllll commented:
There is no perf error with master. I was just demonstrating the difference between a copying approach to get mutability and my non-copying approach. Master does not copy but is also does not give a mutable ndarray. The error would be if you just set `array.flags.writable = True` because then you could do `array.base` and get a `bytes` object which *should* be immutable but now shares memory with a mutable array.

jreback commented:
can u show me a case where we get an exception ? 

I get what u r doing - I am just want to see an example of the practical problem

jreback commented:
I see u did an example way at the top
but I want to see s pandas operation that triggers the error
m

llllllllll commented:
I am not sure when or if I will be able to get to rewriting this in Cython. If someone with more cython experience wants to take this over to help get this done I wouldn't mind and could answer any questions / clarify when the comments are not enough.

llllllllll commented:
ping @jreback

jreback commented on doc/source/whatsnew/v0.18.0.txt:
```diff
@@ -424,6 +424,8 @@ Other enhancements
 - ``HDFStore`` is now iterable: ``for k in store`` is equivalent to ``for k in store.keys()`` (:issue:`12221`).
 - The entire codebase has been ``PEP``-ified (:issue:`12096`)
+- ``from_msgpack`` now always gives writeable ndarrays even when compression is
+  used (:issue:`12359`).
```
move to ``0.18.1`` (other enhancements)

jreback commented on pandas/io/packers.py:
```diff
@@ -255,17 +264,38 @@ def unconvert(values, dtype, compress=None):
-        values = blosc.decompress(values)
-        return np.frombuffer(values, dtype=dtype)
+    if compress:
+        if compress == u'zlib':
```
I would try to import these at the top of the file and just set variables (``_ZLIB_INSTALLED``, ``_BLOSC_INSTALLED``); and you can do whatever imports you need there.

llllllllll replied:
done, added `_check_*` functions to raise a valueerror here if you try to use a compressor that is not installed

jreback commented on setup.py:
```diff
@@ -510,6 +510,13 @@ def pxd(name):
 
 extensions.append(ujson_ext)
 
+# extension for pseudo-safely moving bytes into mutable buffers
```
maybe put this in ``pandas.util._move``?

llllllllll replied:
Okay, there may come a day where there are more general usages for this

llllllllll commented:
Updated based on feedback. I also added standalone tests for `move_into_mutable_buffer` because it is not in `util`.

jreback commented:
IIRC this DID work on windows before.

```
[pandas3] C:\Users\conda\Documents\pandas3.5>python setup.py build_ext --inplace
running build_ext
skipping 'pandas\io/sas\saslib.c' Cython extension (up-to-date)
skipping 'pandas\src\period.c' Cython extension (up-to-date)
skipping 'pandas\hashtable.c' Cython extension (up-to-date)
skipping 'pandas\algos.c' Cython extension (up-to-date)
skipping 'pandas\index.c' Cython extension (up-to-date)
skipping 'pandas\parser.c' Cython extension (up-to-date)
skipping 'pandas\lib.c' Cython extension (up-to-date)
skipping 'pandas\tslib.c' Cython extension (up-to-date)
skipping 'pandas\src\sparse.c' Cython extension (up-to-date)
skipping 'pandas\src\testing.c' Cython extension (up-to-date)
skipping 'pandas\msgpack\_packer.cpp' Cython extension (up-to-date)
skipping 'pandas\msgpack\_unpacker.cpp' Cython extension (up-to-date)
building 'pandas.util._move' extension
creating build\temp.win-amd64-3.5\Release\pandas\util
C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\BIN\amd64\cl.exe /c /nologo /Ox /W3 /GL /DNDEBUG /MD -IC:\Miniconda2\envs\pandas3.5\lib\site-pa
ckages\numpy\core\include -IC:\Miniconda2\envs\pandas3.5\include -IC:\Miniconda2\envs\pandas3.5\include "-IC:\Program Files (x86)\Microsoft Visual Stu
dio 14.0\VC\INCLUDE" "-IC:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\ATLMFC\INCLUDE" "-IC:\Program Files (x86)\Windows Kits\10\include\10.0.
10240.0\ucrt" "-IC:\Program Files (x86)\Windows Kits\NETFXSDK\4.6\include\um" "-IC:\Program Files (x86)\Windows Kits\10\include\10.0.10240.0\shared" "
-IC:\Program Files (x86)\Windows Kits\10\include\10.0.10240.0\um" "-IC:\Program Files (x86)\Windows Kits\10\include\10.0.10240.0\winrt" /Tcpandas/util
/_move.c /Fobuild\temp.win-amd64-3.5\Release\pandas/util/_move.obj
_move.c
pandas/util/_move.c(201): error C2099: initializer is not a constant
error: command 'C:\\Program Files (x86)\\Microsoft Visual Studio 14.0\\VC\\BIN\\amd64\\cl.exe' failed with exit status 2

[pandas3] C:\Users\conda\Documents\pandas3.5>

```

llllllllll commented:
That is saying that my definition for `stolenbuf_type` is not constant but it only uses addesses and int constants, is this using PIC? is there a way to get the compiler to emit the particular field that it does not think is constant?

jreback commented:
https://msdn.microsoft.com/en-us/library/t801az8a.aspx

llllllllll commented:
Read some docs about windows dev and pushed a potential fix. Things still behave normally for me with gcc on gnu+linux, hopefully this builds on windows. Do you have an appveyor build I can check?

jreback commented:
you *could* set this up with appveyor (just create an account), all of the setup will build on your PR's. but its VERY slow. I do with a local vm.

jreback commented on pandas/tests/test_util.py:
```diff
@@ -150,6 +151,38 @@ def test_validation(self):
+class TestMove(tm.TestCase):
+    def test_more_than_one_ref(self):
+        """Test case for when we try to use ``move_into_mutable_buffer`` when
+        the object being moved has other references.
```
its funny, when you use a doc-string it calls the test by this. I guess that is right.

llllllllll replied:
Should I just make this a comment instead or do you not mind?

jreback commented:
ok this passed for me on windows!

merging

jreback commented:
side issue, should prob add some more perf tests with msgpack to cover the bases.

llllllllll commented:
Should I do that in this pr or open another pr with perf tests? Also are those in `test_perf.sh`?

jreback commented:
This is amazing how fast blosc makes this.

no we are using ``asv`` now. (tests are basically in the same place though)

```
In [8]:  %timeit pd.read_msgpack(packed)
100 loops, best of 3: 3.45 ms per loop

In [9]: %timeit df.to_msgpack(compress='blosc')
100 loops, best of 3: 6.27 ms per loop

In [10]: %timeit df.to_msgpack()
100 loops, best of 3: 14.4 ms per loop

In [11]: packed = df.to_msgpack()

In [12]:  %timeit pd.read_msgpack(packed)
10 loops, best of 3: 20.1 ms per loop
```

jreback commented:
@llllllllll thanks!!

jreback commented:
had to make a trivial modification: https://github.com/jreback/pandas/commit/6bba06c87b888aa1bee61a8639d7cfc5bcddc03c

we don't have blosc installed (on purpose) on some of the windows builds. I guess this actually failed travis (but I merged before) :<

jreback commented:
only fails on 2.7 which we were testing with blosc installed. fixed up.

llllllllll commented:
oh, sorry about that. I guess import error is probably more indicative of the problem
